### PR TITLE
Feature/vmo 3976/block error use default exit

### PR DESCRIPTION
**Pull request by AlteaDown**
* Updated the generics of Block/BlockConfig/Prompt/BlockRunners/PromptConfig, which ended up identifying some incorrect types, and will help us avoid issues where a block is not capable of performing a given action. This also results in reduced boilerplate in all call sites, as often, these generics can be ignored, and do not need to be supplied for those cases. Put succinctly, we are making the type system work for us, rather than against us.
* Deprecated findFirstTruthyEvaluatingBlockExitOn until next minor release. Added variants using the `*OrNullI()` and `*OrThrow()` pattern for making these types of calls clearer.
* Removed IBlockExitTestRequired, as it was adding un-necessary complexity, and once removed, didn't affect functionality, as now the "test" field is handled internally in the Block exit functions.
* BlockRunner implementations should be much easier to create now, as all the complexity and edge cases are now handled internally in the IBlock functions, instead of often being forgotten about, and not handled in the BlockRunner implementations.
* Removed broken test that is no longer valid, due to this branch making it so we do not crash, except in extreme cases where there is no error/default exit on a block, and a crash/error is encountered

### CODE REVIEW

File: src/flow-spec/IBlock.ts [hunk 1]
```diff
@@ -107,21 +101,62 @@ export function findBlockExitWith(uuid: string, block: IBlock): IBlockExit {
...
-    exits,
-    ({test, default: isDefault = false}) => !isDefault && evaluateToBool(String(test), evalContext)
-  )
+  return find<IBlockExit>(exits, ({test, default: isDefault = false}) => !isDefault && evaluateToBool(String(test), evalContext))
+}
+
+export function firstTrueBlockExitOrNull(block: IBlock, context: IContext): IBlockExit | undefined {
+  try {
+    return firstTrueBlockExitOrThrow(block, context)
+  } catch (e) {
```
Comment (bulatgab) [inline src/flow-spec/IBlock.ts:122]: Should we check if `e` is `ValidationException` and rethrow if it's not?
(the same goes for other `..OrNull` methods)

File: dist/flow-spec/IBlock.d.ts [hunk 2]
```diff
@@ -1,26 +1,26 @@
-import { IBlockExit, IBlockExitTestRequired, IContext, IContact } from '..';
-export interface IBlock {
+import { IBlockConfig, IBlockExit, IContact, IContext, ISetContactPropertyBlockConfig } from '..';
+export interface IBlock<BLOCK_CONFIG = IBlockConfig, BLOCK_EXIT_CONFIG = {}> {
```
Comment (bulatgab) [inline dist/flow-spec/IBlock.d.ts:2]: ALL_CAPS is an unusual casing for generic parameters, as it's usually used for constants.
But I understand that single letters (like `T`) are not clear enough.

And UpperCamelCase (`BlockConfig` and `BlockExitConfig`) can be confusing - looking like concrete implementations, i.e. class names.
On the other hand, we don't have these classes in this scope, so maybe we could use them (option 1)?

There is also `BlockConfigT` and `GBlockConfig` naming approach which I don't quite like.

And another one is `BlockConfigType` which is not too bad to my taste (option 2), what do you think?

Comment (bulatgab) [reply dist/flow-spec/IBlock.d.ts:2]: Otherwise loved the changes related to the generics!
It was nice to see a number of `any`, `unknown`, and `object` types removed

**Review by bulatgab — APPROVED** [review]

File: src/flow-spec/IBlockExit.ts [hunk 3]
```diff
@@ -78,7 +78,3 @@ export interface IBlockExit {
   // todo: we need to update docs -- they specify "key presence", but I'd prefer us to be more explicit
   default?: boolean
```
Comment (bulatgab) [inline src/flow-spec/IBlockExit.ts:79]: Does our backend use this property? I've tried a tree with some blocks and the exit looked like this:
```
{
  "uuid": "29841127-3247-4610-8000-d26831becf82",
  "label": "29841127-3247-4610-8000-d26831becf82",
  "tag": "Default",
  "destination_block": null,
  "test": "@(TRUE)",
  "config": {}
}
```

Comment (bulatgab) [reply src/flow-spec/IBlockExit.ts:79]: More on the backend changes:
The default exit should probably have `"test": "@(FALSE)"` or simply miss one (undefined) to avoid confusion with regular exits. See the logic in `findDefaultBlockExitOrThrow()` and `_firstBlockExit()`.
Plus the default exit should have `"default": true`.

File: src/flow-spec/IBlock.ts [hunk 4]
```diff
@@ -107,21 +101,62 @@ export function findBlockExitWith(uuid: string, block: IBlock): IBlockExit {
...
+    throw new ValidationException(`All block exits evaluated to false. Block: ${block.uuid}`)
+  }
+  return blockExit
+}
+
+function _firstBlockExit(context: IContext, block: IBlock): IBlockExit | undefined {
+  try {
+    const evalContext = createEvalContextFrom(context)
+    return (
+      find<IBlockExit>(block.exits, blockExit => evaluateToBool(String(blockExit.test), evalContext)) ?? findDefaultBlockExitOnOrNull(block)
```
Comment (bulatgab) [inline src/flow-spec/IBlock.ts:139]: Do we need `?? findDefaultBlockExitOnOrNull(block)`?
Should we `return undefined` instead of `return findDefaultBlockExitOnOrNull(block)` below?

Otherwise, should `firstTrueBlockExitOrThrow` better be called `firstTrueOrDefaultBlockExitOrThrow`?

I feel like we're confusing the default and regular exits here

**Comment by bulatgab** [comment]
A few words on the block exits.

- Shortcomings:

The default exit and the regular exits are stored together in the same array. The order of exits matters - because we’re going from start to end evaluating their `test` property (I think ideally the order shouldn’t matter, there must be a more robust way). There are no types in place to enforce what exits we have in the block. And we must be able to enforce the default exit if the block implies a possibility of an error to be thrown.

- Possible solution:

Maybe the solution here could be to replace the `exits` array with an object and define the default exit as a separate property. Maybe other exits could also be defined as properties (for example, the validation block could have specific exits to handle the validation result), and some - as an array (something like `choiceExits[]` for the SelectOne block, etc.).

File: src/flow-spec/IBlock.ts (hunk 1 above)
Comment (AlteaDown) [reply src/flow-spec/IBlock.ts:122]: For this one, no. The idea is that functions named `*OrNull()` will never throw an error. We are expecting that there can be an error, but we expect null to be returned if that case is encountered. This can lead to much cleaner code, that allows us to decide on the severity of encountering an error at the call site. If we want an error to stop what we are doing, or we want to get information about the error, out of the error object that is thrown, we would use the `*OrThrow()` functions. If we want to just lightly check for an exception, (such as in an if statement's expresssion, where we may be potentially expecting an error case) we can use the `*OrNull()` calls.

File: dist/flow-spec/IBlock.d.ts (hunk 2 above)
Comment (AlteaDown) [reply dist/flow-spec/IBlock.d.ts:2]: The goal is to make it clear that these are not Class names, and they are not standard variables/fields. They are Generics, and especially because of the similar names colliding, it makes sense to make these stand out.

File: src/flow-spec/IBlock.ts (hunk 4 above)
Comment (AlteaDown) [reply src/flow-spec/IBlock.ts:139]: It's necessary, as almost none of the blocks follow the spec, and a block may be marked as `default`, but have no test, and we need to handle that case. We also need to be able to adapt as soon as the block becomes in spec, without breaking anything.
